libs/base/Language/Reflection.idr

```diff
@@ -205,3 +205,14 @@ Elaboration m => MonadTrans t => Monad (t m) => Elaboration (t m) where
   getLocalType        = lift . getLocalType
   getCons             = lift . getCons
   declare             = lift . declare
+
+||| Catch failures and use the `Maybe` monad instead
+export
+catch : Elaboration m => Elab a -> m (Maybe a)
+catch elab = try (Just <$> elab) (pure Nothing)
+
+||| Run proof search to attempt to find a value of the input type.
+||| Useful to check whether a given satisfies an interface constraint.
+export
+search : Elaboration m => (ty : Type) -> m (Maybe ty)
+search ty = catch $ check {expected = ty} `(%search)
```

Comment:
Looks like this is not failing properly: I'm not getting `Nothing` in case the search is unsuccessful
(e.g. `search (Not Nat)`). That leads to `Deriving.Functor` generating improper code.

Reply:
Is `ty` in this case just `Not Nat` or `Prelude.Basics.Not Prelude.Types.Nat`? I mean, I'm not sure `%search` expression does disambiguation and name expansion.

Reply:
Oh, I read your message wrong. What do you mean by "not failing properly"? Does if fail through `Elab` instead of returning `Nothing`?

Reply:
Yes exactly. It fails with "Can't find an implementation for Not Nat" (or something similar) instead of `Nothing`.

Reply:
Found the culprit. And another one too. I have encountered quite a few errors involving "?postpone"
while writing the `derive` script especially with `%runElab f` when `f` has a type of the form
`Elaboration m => m ...` instead of `f : Elab ...`.

So hopefully there will be fewer of these in the future too (the `Search` test case is precisely one such case).

Reply:
Wow! Would this fix #1878 by chance?

Reply:
Nope you still get the hole:
> LOG debug.elab:0: type: ((Builtin.DPair.DPair Prelude.Types.Nat) ?p)

Reply:
I ran into this issue on the current `main` in my complicated script :-( I didn't manage to produce a simple example yet